docs/development/pigs/pig-017.rst

```diff
@@ -0,0 +1,217 @@
+high-level interface. They take advantage of the fact that the ``Analysis`` class has a straight-forward
+access to all the tasks and objects involved in the analysis process, as well as to all the configuration
+settings of the high-level interface. The development of a generic independent Python package that could
+be used for any other Python command-line-tool software and/or API could be explored taking a different
+approach for the capture provenance values but keeping the provenance description file.
+
+There may be some provenance information that could be good to have stuck to the datasets or to other
+objects produced in the data analysis process (i.e. see related discussion in `GH 2447`_, `GH 2452`_
+on accessing and serialising a table of observations metadata and selection cuts together with the derived
+datasets) In those cases the provenance capture and extraction is part of the I/O methods of the object
+where it is linked, and those potential features do not replace the proposal of this PIG.
+
+Alternatives
+============
+
+The are other technical solutions that could be used to capture provenance or perform a structured logging
+in the context of Python scripting executions. The `ctapipe tools`_ come with provenance capture features
+implemented by `ctapipe.core.provenance`_, which capture the execution environment and input/output files.
+Other Python packages that provide provenance capture are `Recipy`_ and `noWorkflow`_. Both provide non-intrusive
+implementation mechanisms and provenance inspection features, but need a relational database (TinyDB, SQLite)
+to store the information and use different different non standard fixed proprietary models. In the end they
+are not highly flexible to configure the kind of information captured and the desired level of granularity.
+
+A different approach is to record provenance as structured logs, which is the one proposed in this PIG. In
+this optic we may find as the most relevant examples `Autologging`_, `Eliot`_ and `Structlog`_. Autologging
+provides an automatic basic logging for class methods calls with parameters values passed and returned, that
+is activated through class decorators. Whereas Eliot and Structlog provide Python APIs to perform a detailed
+structured logging in the form of nested dictionaries and arrays. Eliot also provides a tool to display these
+logs as a structured sequence of nested actions and subactions, as well as the parameters values involved.
+These APIs could be used to develop the custom-model provenance capture layer proposed in this PIG, but they
+add an external dependency that for our simple model we find oversized and could be easily avoided. It remains
+that if technical blockers are found during the development of this PIG, we could always explore the solutions
+offered by these APIs.
+
+Task list
+=========
+
+- General cleaning of Gammapy code-base to replace ``print()`` functions with standard Python ``logging``.
```

Comment:
You can remove this, we already use `logging` consistently instead of `print`. The places where we use `print` is for cases where we explicitly write information to the user screen, e.g. when they print out the information for a catalog source, i.e. someone explicitly calls `print(source_catalogs["hgps"]["crab"])`. Is it clear that those are cases where we in the library should call `print` instead of `log`, or does this need discussion?